Removed the Eigen library as a dependency

Naktakala opened the pull request:
The removal of Eigen as a dependency was straight forward. I also took the chance to port over some math utils from PDT. Functionally nothing really changed except for energy collapsing for TGDSA

Naktakala committed:
Added multiple math utilities for matrix and vector ops

Most of these are ported from PDT

Naktakala committed:
Removal of references to Eigen Library

Naktakala committed:
Merge remote-tracking branch 'upstream/development' into remove_eigen_lib

mhanus commented on CHI_TECH/ChiMath/matrix_operations.cc:
```diff
@@ -0,0 +1,512 @@
+    b = temp;
+  };
+
+  for (unsigned int j = 0; j < AC; j++)
```
No need for the lambda - just use 
```c++
std::swap(A[r1][j], A[r2][j]);
```

Naktakala replied:
Done

mhanus commented on CHI_TECH/ChiMath/matrix_operations.cc:
```diff
@@ -0,0 +1,512 @@
+    b = temp;
+  };
+
+  for (unsigned int i = 0; i < AR; i++)
```
As above - `std::swap`

Naktakala replied:
Done

mhanus commented on CHI_TECH/ChiMath/matrix_operations.cc:
```diff
@@ -0,0 +1,512 @@
+    Ay = MatMul(A, y);
+    lambda = Dot( y, Ay );
+    y = VecMul( Ay, 1./Vec2Norm(Ay) );
+    if (lambda < 0.)
```
The sign flipping could be done after the convergence has been reached, just before returning -- although this is more a simplification of the algorithm rather than a significant performance boost.

Naktakala replied:
Done, although now I needed another fabs.

mhanus commented on CHI_TECH/ChiMesh/SurfaceMesher/Delaunay/delaunay_patch_00_utils.cc:
```diff
@@ -1,5 +1,7 @@
 #include "delaunay_mesher.h"
-#include "Eigen/Dense"
+#include <ChiMath/chi_math.h>
```
```c++
#include "ChiMath/chi_math.h"
```

Naktakala replied:
Done

mhanus commented on CHI_TECH/ChiMath/matrix_operations.cc:
```diff
@@ -0,0 +1,512 @@
+/** Performs power iteration to obtain the fundamental eigen mode. The
+ * eigen-value of the fundamental mode is return whilst the eigen-vector
+ * is return via reference.*/
+double ChiMath::Fundamental_EigValVec_PowerIteration(
```
The name is too complicated IMHO - I think `PowerIteration` would be enough.

Naktakala replied:
Done.

mhanus commented on CHI_TECH/ChiPhysics/PhysicsMaterial/transportxsections/prop_transportxs_03_ecollapse.cc:
```diff
@@ -77,37 +78,38 @@ void chi_physics::TransportCrossSections::
-  Eigen::VectorXd E(G); E.setConstant(1.0);
-  Eigen::VectorXd E_new(G);
 
   //============================================= Perform power iteration
```
Could we use here the power iteration routine from ChiMath?

Naktakala replied:
Done. Good catch actually I thought this was already done.

wdhawkins commented on CHI_TECH/ChiMath/vector_operations.cc:
```diff
@@ -0,0 +1,134 @@
+  assert(y.size() > 0);
+  assert(x.size() == y.size());
+  // Local Variables
+  unsigned int n = x.size();
```
This should be size_t.  Several places in this commit, unsigned int is used with .size().  We should be using size_t instead.

Naktakala replied:
Done. Replaced many references to .size() as unsigned int to size_t

wdhawkins commented:
I'd like to have a discussion about the design of this class.  In particular, I don't see the need for all of these routines to be encapsulated in a class.  I personally would think the chi_math namespace containing individual routines (and typedefs) would be a better design.

Naktakala committed:
Minor changes to math and energy collaps [see description]

Some lambdas have been removed from matrix operations. Sign change logic in power iteration. Swapped power iteration routines in transport energy collapse.

Naktakala committed:
Replaced a lot of size_t references in math.

Naktakala committed:
Moved all matrix operations from ChiMath to chi_math

Naktakala commented:
> I'd like to have a discussion about the design of this class. In particular, I don't see the need for all of these routines to be encapsulated in a class. I personally would think the chi_math namespace containing individual routines (and typedefs) would be a better design.

Done. I moved all matrix and vector operations to chi_math as discussed.
